### PR TITLE
:bug: Avoid data race on package type data

### PR DESCRIPTION
**Pull request by haiyanmeng**
Due to a flawed assumption about dummy packages vs normal packages in
the partial type checker, it was possible to have a data race on thetypecheck data when one package dependended on a given package, and
another had its dependency on that package filtered out.

Makes sure to lock when populating the dummy package data to avoid this.

### CODE REVIEW

File: pkg/loader/loader.go [hunk 1]
```diff
@@ -243,6 +243,11 @@ func (l *loader) typeCheck(pkg *Package) {
 
 		// The imports map is keyed by import path.
 		importedPkg := pkg.Imports()[path]
+
+		// aquire the lock to make sure the access to a given Package is mutual exclusive
```
Comment (mengqiy) [inline pkg/loader/loader.go]: s/aquire/acquire

File: pkg/loader/loader.go [hunk 2]
```diff
@@ -243,6 +243,11 @@ func (l *loader) typeCheck(pkg *Package) {
...
+		importedPkg.Lock()
+		importedPkg.Unlock()
```
Comment (mengqiy) [inline pkg/loader/loader.go]: `defer` is missing.

File: pkg/loader/loader.go (hunk 1 above)
Comment (haiyanmeng) [reply pkg/loader/loader.go]: fixed

File: pkg/loader/loader.go (hunk 2 above)
Comment (haiyanmeng) [reply pkg/loader/loader.go]: fixed.

**Comment by haiyanmeng** [comment]
> I wonder how much performance it may degrade.

Any suggestions on how to measure this efficiently? My gut feeling is that the performance depends on the import structure of a project.
The basic line here is correctness is more important than performance here.

**Comment by mengqiy** [comment]
My initial thought is that the lock covers quite some non-trivial codes, it may degrade the performance. 

But after looking at the travis log you posted above more deeply. It seems it should have been fixed by #241. And I didn't see travis complaining data race on HEAD. Did I miss anything?

**Comment by haiyanmeng** [comment]
> But after looking at the travis log you posted above more deeply. It seems it should have been fixed by #241. And I didn't see travis complaining data race on HEAD. Did I miss anything?

The data race does not happen all the time. 

The race happens between the following two lines:
https://github.com/kubernetes-sigs/controller-tools/blob/0f130d78eeffea1bb30dca6d121ef5360eb10320/pkg/loader/loader.go#L237 
https://github.com/kubernetes-sigs/controller-tools/blob/0f130d78eeffea1bb30dca6d121ef5360eb10320/pkg/loader/loader.go#L250

We already have the lock acquired for the first line:
https://github.com/kubernetes-sigs/controller-tools/blob/0f130d78eeffea1bb30dca6d121ef5360eb10320/pkg/loader/refs.go#L225

So the fix is to add a lock for the second line, which is this PR.

**Comment by mengqiy** [comment]
Another approach in my head is https://github.com/mengqiy/controller-tools/commit/a9371267fb2ff05ff6e6e695c993ad970649c09b
I'm trying to list the options we have before making the changes.
WDYT?

File: pkg/loader/loader.go [hunk 3]
```diff
@@ -243,6 +243,11 @@ func (l *loader) typeCheck(pkg *Package) {
 
 		// The imports map is keyed by import path.
 		importedPkg := pkg.Imports()[path]
+
+		// acquire the lock to make sure the access to a given Package is mutual exclusive
```
Comment (DirectXMan12) [inline pkg/loader/loader.go]: need to mention where the race would be

**Comment by DirectXMan12** [comment]
hmm... something's not right here

**Comment by DirectXMan12** [comment]
ok, so the actual problem is caused by a flaw in the partial type checker's assumptions.  This fix is fine to fix the data race.  The cause is:

given 2 packages, P1 and P2, assume both depend on package C.  Then, assume that P2's reference to  C is filtered out, such that P1's type-checking loop considers C to be a full dependency, and P2's type-checking loop considers C to be a dummy dependency.

Then, it's possible that `P1->check(C)` occurs in parallel with `P2->dummyImport(C)`, since `check(P1)` and `check(P2)` may be called in parallel, and `dummyImport(C)` doesn't currently acquire the package lock.

**Comment by DirectXMan12** [comment]
[migrated race info below to avoid a monster merge commit message]

I noticed this race condition from:
https://travis-ci.org/kubernetes-sigs/controller-tools/jobs/558086207 [see error log 1] and https://travis-ci.org/kubernetes-sigs/controller-tools/jobs/560100634 [see error log 2].

**`Error Log 1`:**
```
••••••••••••==================
WARNING: DATA RACE
Write at 0x00c0002fa198 by goroutine 69:
  sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/loader.go:237 +0x45c
  sigs.k8s.io/controller-tools/pkg/loader.(*Package).NeedTypesInfo()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/loader.go:96 +0xb2
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:249 +0x218
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:243 +0x6c
Previous read at 0x00c0002fa198 by goroutine 70:
  sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck.func1()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/loader.go:250 +0x14d
  sigs.k8s.io/controller-tools/pkg/loader.importerFunc.Import()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/loader.go:349 +0x4b
  go/types.(*Checker).importPackage()
      /home/travis/.gimme/versions/go1.12.linux.amd64/src/go/types/resolver.go:161 +0x8ed
  go/types.(*Checker).collectObjects()
      /home/travis/.gimme/versions/go1.12.linux.amd64/src/go/types/resolver.go:256 +0x2506
  go/types.(*Checker).checkFiles()
      /home/travis/.gimme/versions/go1.12.linux.amd64/src/go/types/check.go:252 +0xa4
  sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck()
      /home/travis/.gimme/versions/go1.12.linux.amd64/src/go/types/check.go:245 +0xa54
  sigs.k8s.io/controller-tools/pkg/loader.(*Package).NeedTypesInfo()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/loader.go:96 +0xb2
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:249 +0x218
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:243 +0x6c
Goroutine 69 (running) created at:
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:241 +0x1e5
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:243 +0x6c
Goroutine 70 (running) created at:
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:241 +0x1e5
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:243 +0x6c
==================
------------------------------
• [SLOW TEST:24.806 seconds]
CRD Generation From Parsing to CustomResourceDefinition
/home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/crd/parser_integration_test.go:38
  should properly generate and flatten the rewritten CronJob schema
  /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/crd/parser_integration_test.go:39
```

**`Error Log 2`:**

```
••••••••••••••••==================
WARNING: DATA RACE
Write at 0x00c0013a8208 by goroutine 71:
  go/types.(*Checker).checkFiles()
      /home/travis/.gimme/versions/go1.12.linux.amd64/src/go/types/check.go:266 +0x14b
  sigs.k8s.io/controller-tools/pkg/loader.(*loader).typeCheck()
      /home/travis/.gimme/versions/go1.12.linux.amd64/src/go/types/check.go:245 +0xa54
  sigs.k8s.io/controller-tools/pkg/loader.(*Package).NeedTypesInfo()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/loader.go:96 +0xb2
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:249 +0x218
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:243 +0x6c
Previous read at 0x00c0013a8208 by goroutine 72:
  [failed to restore the stack]
Goroutine 71 (running) created at:
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:241 +0x1e5
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:243 +0x6c
Goroutine 72 (finished) created at:
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:241 +0x1e5
  sigs.k8s.io/controller-tools/pkg/loader.(*TypeChecker).check.func1()
      /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/loader/refs.go:243 +0x6c
==================
------------------------------
• [SLOW TEST:25.374 seconds]
CRD Generation From Parsing to CustomResourceDefinition
/home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/crd/parser_integration_test.go:38
  should properly generate and flatten the rewritten CronJob schema
  /home/travis/gopath/src/sigs.k8s.io/controller-tools/pkg/crd/parser_integration_test.go:39
------------------------------
```

**Comment by DirectXMan12** [comment]
edited for comment and commit message

/lgtm
/approve

**Pull request merged**